experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/AmActor.scala

```diff
@@ -18,69 +18,113 @@
+  case class AMRMClientAsyncStartup(status:Try[Boolean])
+  case class LaunchContainers(containers: List[Container]) extends YarnApplicationMasterData
+  case class LaunchWorkerContainers(containers: List[Container])
+  case class LaunchServiceContainer(containers: List[Container])
+  case class ContainerRequestMessage(memory: Int, vCores: Int)
+  case class RMHandlerDone(reason: Reason, rMHandlerContainerStats: RMHandlerContainerStats)
+  case class RMHandlerContainerStats(allocated: Int, completed: Int, failed: Int)
+  case class RegisterAMMessage(appHostName: String, appHostPort: Int, appTrackingUrl: String)
+  case class RegisterAppMasterResponse(response: RegisterApplicationMasterResponse)
+  case class AMStatusMessage(appStatus: FinalApplicationStatus, appMessage: String, appTrackingUrl: String)
+  case class ContainerStarted(containerId: ContainerId)
+}
+
+
+class AmActor(appConfig: AppConfig, yarnConf: YarnConfiguration,
+              rmCallbackHandlerActorProps: RMCallbackHandlerActorProps,
+              rmClientActorProps: RMClientActorProps,
+              nmClientAsyncFactory: NMClientAsyncFactory,
+              nodeManagerCallbackHandlerFactory: NodeManagerCallbackHandlerFactory,
+              containerLaunchContextFactory: ContainerLaunchContextFactory) extends Actor {
+  import AmActorProtocol._
+  import AmStates._
 
-/**
- * Yarn ApplicationMaster.
- */
-class AmActor(appConfig: AppConfig, yarnConf: YarnConfiguration) extends Actor {
   val LOG: Logger = LogUtil.getLogger(getClass)
-  val nodeManagerCallbackHandler = createNodeManagerCallbackHandler
-  val nodeManagerClient: NMClientAsync = createNMClient(nodeManagerCallbackHandler)
-  val rmCallbackHandler = context.actorOf(Props(classOf[RMCallbackHandlerActor], appConfig, self), "rmCallbackHandler")
-  val amRMClient = context.actorOf(Props(classOf[ResourceManagerClientActor], yarnConf, self), "amRMClient")
-  val containersStatus = collection.mutable.Map[Long, ContainerInfo]()
+  val nodeManagerCallbackHandler = nodeManagerCallbackHandlerFactory.newInstance(self)
+  val nodeManagerClient: NMClientAsync = nmClientAsyncFactory.newInstance(nodeManagerCallbackHandler, yarnConf)
+  val rmCallbackHandlerActor = context.actorOf(rmCallbackHandlerActorProps.props, "rmCallbackHandler")
+  val rmClientActor = context.actorOf(rmClientActorProps.props, "rmClient")
+
   var masterContainers = Map.empty[ContainerId, (String, Int)]
-  var servicesActor:Option[ActorRef] = None
   val host = InetAddress.getLocalHost.getHostName
   val servicesPort = appConfig.getEnv(SERVICES_PORT).toInt
   val trackingURL = "http://"+host+":"+servicesPort
-
-  var masterAddr:HostPort = _ 
+  var masterAddr: Option[HostPort] = None
   var masterContainersStarted = 0
   var workerContainersStarted = 0
   var workerContainersRequested = 0
-
   val version = appConfig.getEnv("version")
-  
+
+  var servicesActor: Option[ActorRef] = None
+
   override def receive: Receive = {
```

Comment:
Better to break this receive handler to several handler, so that it is more clear what the case branch is actually doing.